lib/private/RedisFactory.php

```
@@ -150,12 +155,44 @@ private function getSslContext($config) {
 		return null;
 	}
 
-	public function getInstance() {
+	public function getInstance(bool $retry = true) {
 		if (!$this->isAvailable()) {
 			throw new \Exception('Redis support is not available');
 		}
-		if (!$this->instance instanceof \Redis) {
+
+		if ($this->instance !== null) {
+			return $this->instance;
+		}
+
+		try {
+			// REDIS_CONNECTION_COOLDOWN seconds absolute cooldown before trying to reconnect in a different request
+			$cooldown = $this->config->getValue('redis.conn_cooldown', 0);
+			if (\time() - $cooldown < self::REDIS_CONNECTION_COOLDOWN) {
+				$this->logger->info('Redis connection is in cooldown, skipping connection attempt');
+				$this->instance = null;
+				return null;
+			}
+
 			$this->create();
+			$this->config->deleteValue('redis.conn_cooldown');
+		} catch (\RedisException $e) {
+			$this->instance = null;
+
+			if ($retry) {
+				// Retry once
+				$this->logger->info(
+					'Failed to connect to Redis, retrying once before entering cooldown',
+					['app' => 'core', 'exception' => $e],
+				);
+				usleep(300000); // 300ms
+				return $this->getInstance(false);
+			}
+
+			$this->logger->error(
+				'Failed to connect to Redis, entering cooldown of ' . self::REDIS_CONNECTION_COOLDOWN . ' seconds before retrying',
+				['app' => 'core', 'exception' => $e],
+			);
+			$this->config->setValue('redis.conn_cooldown', \time());
```

This will fail hard as well if the config is not writable. Similarly it will be an issue if numerous requests try to write the config file at the same time.